text/0034-sdk-lifecycle-hooks.md

```diff
@@ -0,0 +1,125 @@
+
+To test out this RFC, we implemented it in the [JavaScript SDK](https://github.com/getsentry/sentry-javascript/blob/7aa20d04a3d61f30600ed6367ca7151d183a8fc9/packages/types/src/client.ts#L153) with great success, so now we are looking to propose and implement it in the other SDKs.
+
+Lifecyle hooks can be registered on a Sentry client, and allow for integrations/sdk users to have finely grained control over the event lifecycle in the SDK.
+
+```ts
+interface Client {
+  // ...
+
+  on(hookName: string, callback: (...args: unknown) => void) => void;
+
+  emit(hookName: string, ...args: unknown[]) => void;
+
+  // ...
+}
+```
+
+# Motivation
+
+There are three main ways users can extend functionality in the SDKs right now.
+
+At it's current form, the SDK is an event processing pipeline. It takes in some data (an error/message, a span, a profile), turns it into the event, attaches useful context to that event based on the current scope, and then sends that event to Sentry.
+
+```
+| Error | ---> | Event | ---> | EventWithContext | ---> | Envelope | ---> | Transport | ---> | Sentry |
+```
+
+```
+| TransactionStart | ---> | SpanStart | ---> | SpanFinish | ---> | TransactionFinish | --> | Event | ---> | EventWithContext | ---> | Envelope | ---> | Transport | ---> | Sentry |
+```
+
+```
+| Session | ---> | Envelope | ---> | Transport | ---> | Sentry |
+```
+
+The SDKs provide a few ways to extend this pipeline:
+
+1. Event Processors (what Integrations use)
+2. `beforeSend` callback
+3. `beforeBreadcrumb` callback
+
+But these are all top level options in someway, and are part of the unified API as a result. This means that in certain scenarios, they are not granular enough as extension points.
+
+The following are some examples of how sdk hooks can unblock new features and integrations, but not a definitive list:
+
+- Integrations want to add information to spans when they start or finish. This is what [RFC #75 is running into](https://github.com/getsentry/rfcs/pull/75), where they want to add thread information to each span.
+- Integrations want to add information to envelopes before they are sent to Sentry.
+- Integrations want to run code on transaction/span finish (to add additional spans to the transaction, for example).
+- Integrations want to mutate an error on `captureException`
+- Integrations want to override propagation behaviour (extracing/injecting outgoing headers)
+
+# Proposal
+
+SDK hooks live on the client, and are **stateless**. They are called in the order they are registered. SDKs can opt-in to whatever hooks they use, and there can be hooks unique to an SDK.
+
+Hooks are meant to be mostly internal APIs for integration authors, but we can also expose them to SDK users if there is a use case for it.
+
+As hook callbacks are not processed by the client, they can be async functions.
+
+```ts
+// Example implementation in JavaScript
+
+type HookCallback = (...args: unknown[]): void;
+
+class Client {
```

Comment:
I am a bit confused, who implements this `Client`? What exactly the end-user interacts with?

Reply:
Thanks a lot for showing an interset in our RFC process! 🚀 
This is meant to be the client implementation we have in each SDK, see https://develop.sentry.dev/sdk/unified-api/#client

Reply:
I need to think a layer higher and across multiple languages and clients. Sorry for the confusion, and thank you so much for your response!

Reply:
@cleptric, I also need some clarification. Does this mean I need to implement a client when I want to add lifecycle hooks? Can you share the code a user must write when implementing a lifecycle hook?

Reply:
The `on` and `emit` hooks will be defined on the client interface (much like `flush` and `capture_event` are at the moment).

Usually users (if writing custom clients) are inheriting from a base client implementation anyway, so they should get the hooks functionality out of the box.